File: app/src/main/java/bogomolov/aa/fitrack/viewmodels/TagSelectionViewModel.java
package bogomolov.aa.fitrack.viewmodels;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import bogomolov.aa.fitrack.repository.Repository;
import bogomolov.aa.fitrack.core.model.Tag;

import static bogomolov.aa.fitrack.core.Rx.*;


public class TagSelectionViewModel extends ViewModel {
    public MutableLiveData<String> newName = new MutableLiveData<>();
    public MutableLiveData<List<Tag>> tagsLiveData = new MutableLiveData<>();
    private Repository repository;
    private List<Tag> tags;

    @Inject
    public TagSelectionViewModel(Repository repository) {
        this.repository = repository;
        worker(()->{
            tags = repository.getTags();
            tagsLiveData.postValue(tags);
        });
    }

    public void onNewTag() {
        worker(()->{
            String tagName = newName.getValue();
            Tag tag = new Tag(tagName);
            repository.addTag(tag);
            tags.add(tag);
            tagsLiveData.postValue(tags);
        });
    }

    public void deleteTag(Tag tag) {
        worker(()->{
            repository.deleteTag(tag);
            tags.remove(tag);
            tagsLiveData.postValue(tags);
        });
    }
}

File: app/src/main/java/bogomolov/aa/fitrack/viewmodels/TagSelectionViewModel.kt
package bogomolov.aa.fitrack.viewmodels

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope

import javax.inject.Inject

import bogomolov.aa.fitrack.repository.Repository
import bogomolov.aa.fitrack.core.model.Tag

import bogomolov.aa.fitrack.android.worker


class TagSelectionViewModel @Inject
constructor(private val repository: Repository) : ViewModel() {
    var newName = MutableLiveData<String>()
    var tagsLiveData = MutableLiveData<List<Tag>>()
    private var tags: MutableList<Tag>? = null

    init {
        worker(viewModelScope) {
            tags = repository.getTags() as MutableList<Tag>?
            tagsLiveData.postValue(tags)
        }
    }

    fun onNewTag() {
        worker(viewModelScope) {
            val tagName = newName.value
            if (tagName != null) {
                val tag = Tag(tagName)
                repository.addTag(tag)
                tags!!.add(tag)
                tagsLiveData.postValue(tags)
            }
        }
    }

    fun deleteTag(tag: Tag) {
        worker(viewModelScope) {
            repository.deleteTag(tag)
            tags!!.remove(tag)
            tagsLiveData.postValue(tags)
        }
    }
}